R/transformCounts.R

```diff
@@ -779,6 +793,52 @@ setMethod("transformAssay", signature = c(x = "SingleCellExperiment"),
     return(mat)
 }
 
+########################### .apply_transformation_difference ##################
+# Computes all pairwise differences (x - y) between features across samples.
+# Returns a matrix with one row per feature pair.
+.apply_transformation_difference <- function(mat, ...) {
+  # Get the feature (taxa) names
+  taxa <- rownames(mat)
+  
+  # Generate all unique pairwise combinations of taxa
+  combs <- utils::combn(taxa, 2, simplify = FALSE)
+  
+  # Compute pairwise differences: x - y for each taxa pair across samples
+  diffs <- lapply(combs, function(pair) {
+    t1 <- pair[1]; t2 <- pair[2]
+    name <- paste0("diff_", t1, "_", t2)
+    values <- mat[t1, ] - mat[t2, ]
+    matrix(values, nrow = 1, dimnames = list(name, colnames(mat)))
+  })
+  
+  # Combine all difference rows into a single matrix
+  result <- do.call(rbind, diffs)
+  
+  # Add metadata
+  attr(result, "mia") <- "diff"
+  
+  return(result)
+}
+
+############################ .apply_transformation_division ############################
+# Computes all pairwise ratios (x / y) between features across samples.
+# Returns a matrix with one row per feature pair.
+.apply_transformation_division <- function(mat, ...) {
+  taxa <- rownames(mat)
+  combs <- utils::combn(taxa, 2, simplify = FALSE)
+  
+  ratios <- lapply(combs, function(pair) {
+    t1 <- pair[1]; t2 <- pair[2]
+    name <- paste0("div_", t1, "_over_", t2)
+    values <- mat[t1, ] / mat[t2, ]
```

Comment:
What happens if there are zeroes in mat[t2,] - is this tested and is pseudocount supported? (perhaps it has support already in the upstream before we get in this point?). At least an informative warning could be thrown here.